### PR TITLE
fixed function parameter reflection

### PR DESCRIPTION
**Pull request by vlastavesely**
```
ApiGen\Exception\Templating\FailedRenderFileException: Rendering of "/home/*/apigen/packages/ThemeDefault/src/function.latte" template to "/tmp/_apigen_cache/function-ApiGen.Tests.Generator.Source.someOtherFunction.html" file failed due to: Return value of ApiGen\Reflection\Reflection\AbstractParameterReflection::getAnnotation() must be an instance of phpDocumentor\Reflection\DocBlock\Tags\Param or null, instance of phpDocumentor\Reflection\DocBlock\Tags\Return_ returned
```

```
ApiGen\Exception\Templating\FailedRenderFileException: Rendering of "/home/*/apigen/packages/ThemeDefault/src/function.latte" template to "/tmp/_apigen_cache/function-ApiGen.Tests.Generator.Source.someOtherFunction.html" file failed due to: Call to undefined method ApiGen\Reflection\Reflection\Function_\FunctionParameterReflection::isDefaultValueAvailable()
```

```
ApiGen\Exception\Templating\FailedRenderFileException: Rendering of "/home/*/apigen/packages/ThemeDefault/src/function.latte" template to "/tmp/_apigen_cache/function-ApiGen.Tests.Generator.Source.someOtherFunction.html" file failed due to: Argument 1 passed to ApiGen\Annotation\Latte\Filter\AnnotationFilterProvider::ApiGen\Annotation\Latte\Filter\{closure}() must implement interface ApiGen\Reflection\Contract\Reflection\Partial\AnnotationsInterface, instance of ApiGen\Reflection\Reflection\Function_\FunctionParameterReflection given, called in /tmp/_latte_cache/ThemeDefault-src-function.latte--2096778438.php on line 154
```

### CODE REVIEW

**Comment by TomasVotruba** [comment]
Thanks for the PR. Could you please more description? It's difficult for me to read your intention based on 3 errors.

File: packages/ThemeDefault/src/function.latte [hunk 1]
```diff
@@ -51,7 +51,7 @@
                     <span class="property-name">{if $parameter->isPassedByReference()}&amp; {/if}${$parameter->getName()}</span>
                     {if $parameter->isDefaultValueAvailable()} = {$parameter->getDefaultValueDefinition()|highlightValue:$function|noescape}{elseif $parameter->isVariadic()},…{/if}
                 {/block}</code></td>
-                <td>{$parameter|description}</td>
+                <td>{$parameter->getDescription()}</td>
```
Comment (TomasVotruba) [inline packages/ThemeDefault/src/function.latte]: Why is this needed? It would disable possibility to tune description via event. That's why there is filter "description" here, that calls `->getDescription()` internally.

File: tests/Generator/Source/SomeFunction.php [hunk 2]
```diff
@@ -6,6 +6,9 @@ function someFunction(): void
 {
 }
 
-function someOtherFunction(): void
+/**
+ * @return void
+ */
+function someOtherFunction($paramWithoutTypeHint): void
```
Comment (TomasVotruba) [inline tests/Generator/Source/SomeFunction.php]: Could you keep both types of functions - returning `: void`, one with annotation, second without?

So both cases are covered with test.

File: packages/Reflection/src/Reflection/AbstractParameterReflection.php [hunk 3]
```diff
@@ -67,6 +67,11 @@ public function getTypeHintClassOrInterfaceReflection()
         return $classOrInterfaceReflection;
     }
 
+    public function isDefaultValueAvailable(): bool
```
Comment (TomasVotruba) [inline packages/Reflection/src/Reflection/AbstractParameterReflection.php]: Could please add this method to `ParameterReflectionInterface`?

Comment (vlastavesely) [inline packages/Reflection/src/Reflection/AbstractParameterReflection.php]: you mean `AbstractParameterReflectionInterface`?

File: packages/ThemeDefault/src/function.latte (hunk 1 above)
Comment (vlastavesely) [inline packages/ThemeDefault/src/function.latte]: yes, it is mistake. in the evening I watched some of your commits and realized that this is not right. I am going to edit it.

Where latte filters are defined?

**Comment by vlastavesely** [comment]
It could happen that a parameter of a function does not have anotation. If if happens, there is not returned instance of `AnnotationsInterface` and latte filter fails because of it.

Separated problem was missing `AbstractParameterReflection::isDefaultValueAvailable()`.

And third problem was that function `AbstractParameterReflection::getAnnotation()` could return annotation for `@return` instead `@param`.

File: packages/Reflection/src/Reflection/AbstractParameterReflection.php (hunk 3 above)
Comment (TomasVotruba) [inline packages/Reflection/src/Reflection/AbstractParameterReflection.php]: Yep

**Comment by TomasVotruba** [comment]
I see, now I get it. Thanks for expalining!

**Comment by TomasVotruba** [comment]
As for `Return`/`Param` annotation, maybe (there should be)[$annotations = $declaringReflection->getAnnotations()]:

```php
$annotations = $declaringReflection->getAnnotation(AnnotationList::PARAM);
```

That should prevent both `instanceof Param` checks bellow and in a template.

**Comment by vlastavesely** [comment]
yes, this is better. I was not aware of fact that function getAnnotations() can take an argument. :)

**Comment by TomasVotruba** [comment]
@vlastavesely Good point. Maybe `getAnnotationsByType()` would be better.

**Comment by vlastavesely** [comment]
In template, problem persists:
```
ApiGen\Exception\Templating\FailedRenderFileException: Rendering of "/home/*/apigen/packages/ThemeDefault/src/function.latte" template to "/tmp/_apigen_cache/function-ApiGen.Tests.Generator.Source.someOtherFunction.html" file failed due to: Argument 1 passed to ApiGen\Annotation\Latte\Filter\AnnotationFilterProvider::ApiGen\Annotation\Latte\Filter\{closure}() must implement interface ApiGen\Reflection\Contract\Reflection\Partial\AnnotationsInterface, instance of ApiGen\Reflection\Reflection\Function_\FunctionParameterReflection given, called in /tmp/_latte_cache/ThemeDefault-src-function.latte--2096778438.php on line 154
```

**Comment by TomasVotruba** [comment]
I see. Let's have it this way for now. Is it ready to merge?

**Comment by vlastavesely** [comment]
Only thing that remains, would be method in class and in trait - I am afraid that it could cause same problem. Give me a few minutes.

**Comment by TomasVotruba** [comment]
Allright, no need to hurry, I was just curious. Take your time

**Comment by vlastavesely** [comment]
I tried it and it was a little bit buggy.

Main new thing: in `method.latte` there are information about overriden method and what interfaces' method is implemented. But trait cannot extend or implement anything. So now it is shown just or class.

**Comment by TomasVotruba** [comment]
I'll look on later.

Could you fix cs as shown in Travis?

**Comment by vlastavesely** [comment]
I already fixed. And then realized that my PR is dumplicate to #909 :)

**Comment by TomasVotruba** [comment]
I see, great. You're so fast I can't keep up on my mobile phone :)

**Comment by TomasVotruba** [comment]
I see, great. You're so fast I can't keep up on my mobile phone :)

**Comment by TomasVotruba** [comment]
Thank you for the the fix and compleeting missing methods!
